lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5122,6 +5138,19 @@ static void diagnoseClassWithoutInitializers(TypeChecker &tc,
 }
 
 void TypeChecker::maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+  if (auto *SF = classDecl->getParentSourceFile()) {
+    // Allow classes without initializers in SIL and textual interface files.
```

Comment:
There are a lot of these switches over `SourceFileKind`. Can they be grouped into some smaller number of predicates, e.g., `declarationRequiresDefinition(decl)` could cover both function bodies and variables needing initializers, so that code doesn't need to be duplicated.

Reply:
We could, but I'm not sure they're always the same cases. *Usually* it's SIL and textual interfaces that allow some part of the declaration to be missing, but in some cases SIL cares too, and in one case (initial values for globals) script mode files are also exempt. That made me hesitant to factor it out into anything general. They could all be predicates on the various individual Decls, though.